controllers/external/util.go

```diff
@@ -25,23 +25,24 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```

Comment:
@Danil-Grigorev Did some analysis (sharing here for future-us :))

Ref fields with namespace defaulting before this PR:
* Cluster.Spec.InfrastructureRef
* Cluster.Spec.ControlPlaneRef
* Machine.Spec.InfrastructureRef
* Machine.Spec.Bootstrap.ConfigRef
* MachinePool.Spec.Template.Spec.Bootstrap.ConfigRef
* MachinePool.Spec.Template.Spec.InfrastructureRef
* MHC.Spec.RemediationTemplate
* KCP.spec.machineTemplate.infrastructureRef
* ClusterClass (all refs)

Namespace fields for which we have to add defaulting:
* MachineDeployment (not used anymore for rollout decision, namespace is set by Cluster topology controller)
  * [x] Defaulting already added with this PR
  * [ ] DeepCopy ref and set namespace if empty directly before external.Get (in reconcileExternalTemplateReference)
* MachineSet
  * [x] Defaulting already added with this PR
  * [ ] DeepCopy ref and set namespace if empty directly before external.Get (in reconcileExternalTemplateReference)

So I think we're almost good. I think we just can't assume that the defaulting/mutating webhook always run before our controllers (e.g. if a MD/MS is never modified). I think for this case it would be good to add safeguards in the reconcileExternalTemplateReference funcs to set the namespace if it is empty directly before we use it (but after UpdateReferenceAPIContract))